WooCommerce/Classes/Extensions/String+Helpers.swift

```diff
@@ -68,4 +68,150 @@ extension String {
+                outString += String(characterAtIndex)
+            }
+
+            return outString
+        }
+
+        /// Ranked per https://www.php.net/manual/en/function.version-compare.php
+        ///
+        enum ComponentScore: Comparable {
+            case other
+            case dev
+            case alpha
+            case beta
+            case RC
+            case number
+            case patch
+        }
+
+        /// Score each component
+        ///
+        func scoreComponent(stringComponent: String) -> ComponentScore {
+            if stringComponent == "dev" {
+                return .dev
+            }
+            if stringComponent == "alpha" || stringComponent == "a" {
+                return .alpha
+            }
+            if stringComponent == "beta" || stringComponent == "b" {
+                return .beta
+            }
+            if stringComponent == "RC" || stringComponent == "rc" {
+                return .RC
+            }
+            let componentCharacterSet = CharacterSet(charactersIn: stringComponent)
+            if componentCharacterSet.isSubset(of: .decimalDigits) {
+                return .number
+            }
+
+            if stringComponent == "pl" || stringComponent == "p" {
+                return .patch
+            }
+
+            return .other
+        }
+
+        /// Score and compare two string components
+        ///
+        func compareStringComponent(stringComponent1: String, stringComponent2: String) -> ComparisonResult {
+            /// Score each component
+            let stringComponent1Score = scoreComponent(stringComponent: stringComponent1)
+            let stringComponent2Score = scoreComponent(stringComponent: stringComponent2)
+
+            if stringComponent1Score < stringComponent2Score {
+                return .orderedAscending
+            }
+
+            if stringComponent1Score > stringComponent2Score {
+                return .orderedDescending
+            }
+
+            if stringComponent1Score == .number && stringComponent2Score == .number {
+                let component1Int = Int(stringComponent1) ?? 0
+                let component2Int = Int(stringComponent2) ?? 0
+
+                if component1Int < component2Int {
+                    return .orderedAscending
+                }
+
+                if component1Int > component2Int {
+                    return .orderedDescending
+                }
+            }
+
+            return .orderedSame
+        }
+
+        var leftHandString = replaceUnderscoreDashAndPlusWithDot(inString: self)
+        leftHandString = insertDotsBeforeAndAfterAnyNonNumber(inString: leftHandString)
+
+        var rightHandString = replaceUnderscoreDashAndPlusWithDot(inString: to)
+        rightHandString = insertDotsBeforeAndAfterAnyNonNumber(inString: rightHandString)
+
+        let leftComponents = leftHandString.components(separatedBy: ".")
+        let rightComponents = rightHandString.components(separatedBy: ".")
```

Comment:
nit: the two transformations on the string could probably be done in another helper function for reuse, like:

```swift
func versionComponents(of string: String) -> [String] {
    // Call `replaceUnderscoreDashAndPlusWithDot`
    // Call `insertDotsBeforeAndAfterAnyNonNumber`
    // Return components
}
```

Another idea is to update `replaceUnderscoreDashAndPlusWithDot` and `insertDotsBeforeAndAfterAnyNonNumber` to private `String` extension functions, so that you can just chain the calls like:

```swift
func versionComponents(of string: String) -> [String] {
    string.replacingUnderscoreDashAndPlusWithDot().insertingDotsBeforeAndAfterAnyNonNumber().components(separatedBy: ".")
}
```

Reply:
DRYed the pre-processing of the strings in https://github.com/woocommerce/woocommerce-ios/pull/5382/commits/2aabed7489bd190cf2bc7facd80eae989ff8e154

I'm going to pass on the chaining suggestion, but thank you! I prefer seeing the transformations on separate lines. OK?